reconcile/terraform_resources.py

```diff
@@ -415,8 +417,11 @@ def fetch_current_state(dry_run, namespaces, thread_pool_size,
     return ri, oc_map
 
 
-def init_working_dirs(accounts, thread_pool_size,
```

Comment:
According to [style guide ](https://service.pages.redhat.com/dev-guidelines/docs/appsre/python-style-guide/#abstract-collection-types) this should be `Iterable[Mapping[str, Any]]`?

Reply:
This was done on purpose. We have some code that takes an input, say a `dict` and returns that object directly (rather than creating a new data structure). Doing this makes it difficult to use abstract types for parameters and concrete for return.

For an example, see `dump` where we're doing:

```python
    def dump(self, print_to_file=None, existing_dirs=None):
        if existing_dirs is None:
            working_dirs = {}
        else:
            working_dirs = existing_dirs

       ....

       return working_dirs
```

In this scenario, the return can either be a `dict`, or whatever we passed into `existing_dirs`. If I restrict `existing_dirs` to a `Mapping`, then the return value can either be a `Mapping` or `dict`, so I can't set the return value to `dict`. I'd really have to change how the code works to fix this, which I don't want to do given the limited coverage here. I only broke that rule twice for a couple of `cast` calls which return the data unmodified.

Anyone writing this function today with type hints would discover this issue and then probably call `dict(existing_dirs)` to cast the input to `dict`, but unless we feel strongly about this, I'd rather keep any actual changes to how the code could behave out of this PR. I think this would be best fixed in a separate PR with the addition of tests (if that is practical).

Reply:
Thx a lot for the detailed explanation! :bow: Makes sense now